### PR TITLE
Add flag to create a bundle per test file

### PR DESCRIPTION
**Pull request by marvinhagemeister**
This PR brings back an option to control whether we want to bundle all files into a single bundle or create a bundle per file. By setting `singleBundle: false` we'll create a bundle for each test file. The default value is `true`

### CODE REVIEW

File: src/bundle.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import * as path from "path";
 import * as esbuild from "esbuild";
-import { Deferred, formatTime } from "./utils";
+import { EventEmitter } from "events";
```
Comment (jridgewell) [inline src/bundle.ts:3]:
```suggestion
import type { EventEmitter } from "events";
```

File: src/bundler-map.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import { EventEmitter } from "events";
```
Comment (jridgewell) [inline src/bundler-map.ts:1]:
```suggestion
import type { EventEmitter } from "events";
```

File: src/bundle.ts [hunk 3]
```diff
@@ -33,32 +30,24 @@ export class Bundle {
...
+		private emitter: EventEmitter,
+	) {
+		this.config = { ...config, entryPoints: [file] };
 	}
 
 	dirty() {
 		if (this._dirty) return;
 		this._dirty = true;
 		this.deferred = new Deferred();
+		if (this.buildInProgress) this.write();
```
Comment (jridgewell) [inline src/bundle.ts:46]: I don't remember why this was necessary anymore. Can you delete and see if one of the tests fails?

It seems the rewriting of `this.deferred` above will ensure any waiters on the old promise will resolve only once the new build happens, and that should happen with the next normal write.

File: src/index.ts [hunk 4]
```diff
@@ -108,54 +116,102 @@ function createPreprocessor(
...
+		const match = /^\/(?:absolute|base\/)([^?#]*?)(\.map)?(?:\?|#|$)/.exec(
+			req.url || "",
+		);
+		if (!match) {
+			return next();
+		}
 
-		const filePath = path.normalize(match[1]);
-		if (filePath !== bundle.file) return next();
+		const absolutePath = path.isAbsolute(match[1])
```
Comment (jridgewell) [inline src/index.ts:208]: I believe this could just be done by by checking to see if the url started with `/base/`? We could turn `(?:absolute|base\/)` into `(absolute|base\/)` and then check the first match, too.

But why is this necessary? Shouldn't all paths be `/absolute`?

File: src/index.ts [hunk 5]
```diff
@@ -52,21 +59,25 @@ function createPreprocessor(
...
+	if (singleBundle) {
+		// Push the entry point so that Karma will load the file when the runner starts.
+		config.files.push({
+			pattern: testEntryPoint.file,
+			included: true,
+			served: false,
+			watched: false,
+			type: format === "esm" ? "module" : "js",
+		});
+		testEntryPoint.touch();
```
Comment (jridgewell) [inline src/index.ts:72]: Nit: We could also construct the `testEntryPoint` here. It's not needed anywhere else but this function.

File: src/index.ts [hunk 6]
```diff
@@ -108,54 +116,102 @@ function createPreprocessor(
...
+			log.info(`Compiling to ${ev.file}...`);
+		} else if (!pendingBundles.size) {
+			start = ev.time;
+			log.info(`Compiling...`);
+		}
+		pendingBundles.add(ev.file);
+	});
+
+	bundlerMap.on("stop", ev => {
+		pendingBundles.delete(ev.file);
```
Comment (jridgewell) [inline src/index.ts:137]: I think we can just `clear` here? Stopping stops everything. We might even need this event, the karma process will die shortly.

File: src/index.ts [hunk 7]
```diff
@@ -108,54 +116,102 @@ function createPreprocessor(
...
+	});
+
+	bundlerMap.on("done", ev => {
+		pendingBundles.delete(ev.file);
+		if (singleBundle || pendingBundles.size === 0) {
+			log.info(`Compiling done (${formatTime(ev.endTime - start)})`);
+		}
 	});
 
 	const buildBundle = debounce(() => {
```
Comment (jridgewell) [inline src/index.ts:147]: We should rename this to `buildSingleBundle`, since it's only called in `singleBundle` mode.

File: src/index.ts [hunk 8]
```diff
@@ -108,54 +116,102 @@ function createPreprocessor(
...
+			await buildBundle();
+			// Turn the file into a `dom` type with empty contents to get Karma to
+			// inject the contents as HTML text. Since the contents are empty, it
+			// effectively drops the script from being included into the Karma runner.
+			file.type = "dom";
+			done(null, "");
+		} else {
+			const res = await bundlerMap.read(filePath);
+
+			(file as any).sourceMap = res.map;
```
Comment (jridgewell) [inline src/index.ts:182]: This might not be necessary with the `formatError` function. Or, we could skip installing `formatError` in `singleBundle` mode?

**Review by jridgewell — COMMENTED** [review]
A few comments (mostly nitty). Mostly looks good.

File: src/bundle.ts (hunk 3 above)
Comment (marvinhagemeister) [reply src/bundle.ts:46]: Good call this can be removed. All tests are passing without this line 👍

File: src/index.ts (hunk 5 above)
Comment (marvinhagemeister) [reply src/index.ts:75]: Tried that, but it would require adding a quite a few type checks, as TS complains that it may be undefined. Imo it's ok if we leave it as is.

File: test/fixtures/bundle-per-file/files/dep1.js [hunk 9]
```diff
@@ -0,0 +1,3 @@
+export function foo() {
+	return "CONTENT";
```
[the remark this answers is not in the document]
Comment (jridgewell) [reply test/fixtures/bundle-per-file/files/dep1.js:2]: It seems this suggestion was missed.

**Review by jridgewell — COMMENTED** [review]
Belated LGTM.

File: test/fixtures/bundle-per-file/files/dep1.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (marvinhagemeister) [reply test/fixtures/bundle-per-file/files/dep1.js:2]: Shoot, I had that change, must have lost it during rebase
